feat(enhancement): Add "timer" trigger for missions

danaris opened the pull request:
**Feature:** This PR implements the feature request detailed and discussed in issue #4475 

Adds a new `timer` element to mission specifications, allowing for actions to trigger after a certain amount of time spent in a given system.

## Feature Details
The element has the following format:

`timer "<Timer Name>" <base time> [random time]`
`    system "<system name>"`
`    [idle]`
`    [uncloaked]`
`    [proximity <radius> [close|far]]`
`        ["<planet name>"]`
`    on timeup`
`        <action/conversation>`

## UI Screenshots
N/A

## Usage Examples
This PR includes a modification of the "Sad Archie" mission to use the timer rather than the existing Timer Ship mechanism.

## Testing Done
- [x] the timer correctly loads from and saves to the datafiles
- [x] it times the number of seconds specified
- [x] it pauses on activity, on cloak, and when the player is too far away from the target zone
- [x] it correctly triggers the action
- [x] it correctly saves the fact that the timer completed to the savefile
- [x] it correctly allows the mission to complete when the timer is complete
- [ ] it saves the time remaining when leaving the system if allowed
- [ ] it saves the time remaining when landing, quitting, and reloading the save if allowed
- [ ] it resets the time remaining on leaving the target zone, on leaving the system, or on any action that pauses the countdown if specified
- [ ] it safely handles the target system being removed from the map after the mission is loaded
- [ ] it correctly prevents the mission from completing when the timer is not yet complete

### Automated Tests Added
I have no experience with creating tests, and am unclear on whether this needs them.

## Performance Impact
Performance impact should be minimal. It adds a loop through active missions to check for timers, and through the active timers to step through them, but each call of the `Timer::Step()` method is quite lightweight, and no appreciable performance impact was observed.

danaris commented:
As I was playing through the Wanderer campaign, I noticed another place where it seemed like the `timer` functionality would be very useful: in the "Wanderers: Kor Mereti Hacking" mission.

I changed the `on enter Mekislepti` trigger to the following timers:

```
	timer "Kor Mereti Hacking Timer A" 5
		system Mekislepti
		uncloaked
		reset "leave system"
		on timeup
			"reputation: Kor Mereti" = 1
	
	timer "Kor Mereti Hacking Timer B" 10
		system Mekislepti
		uncloaked
		on timeup
			"reputation: Kor Mereti" = -10
			dialog
				`When you enter the Mekislepti system, an entire fleet of drones moves to intercept you. Hardly daring to hope, you press the button that activates the Wanderer program. For a fraction of a second, the drones' engines stop firing and they drift as if out of control. Then, with renewed purpose, they move to attack you.`
				`	Oh well. It was worth a shot.`
```

...and the result produced [this video](http://topazgryphon.org/~tcollett/mereti_hacking_timer.mp4).

It does make me think we might need another flag for timers: one that prevents them from stopping or resetting for any reason. Without something like that, there would be a danger, in this case, of the player leaving the system within the 5 seconds that the Kor Mereti are friendly; I don't know if that would be useful, but it is at least a concern.

quyykk commented on data/sheragi/archaeology missions.txt:
```diff
@@ -345,17 +345,18 @@ mission "Sad Archie"
-			`	Time speeds up, and you watch as the dragon cities grow and spread out. You see one city buried under a massive lava flow, starships buzzing around it like a cloud of flies to carry people and goods to safety. Other cities are destroyed by war, and then abruptly the dragon-people are all gone and their cities lie vacant. Buildings crumble, and the forests reclaim the land.`
-			`	Then you wake up and find yourself back in your ship, in the Ildaria system. Weird.`
+	timer "Sad Archie Timer" 120 60
+		system Ildaria
```
Should this be a location filter instead?

danaris replied:
For this specific mission, definitely not; it's explicitly supposed to be Ildaria.

In general...I considered it, but it seems to me that if you're giving a player an invisible timer, it's at least polite to make sure they know _exactly_ where they have to wait for it, rather than expecting them to guess one of N systems.

I would think that, if it _were_ to support location filters, there would need to be a way to get the mission text to recognize which system was selected, so it could pass that information along. This is particularly tricky if a mission has more than one timer (which is supported right now). Either that, or the location filter would have to be on the mission itself, and there would have to be a way of passing that information through to the timer.

Unless I'm missing something, which is entirely possible.

danaris replied:
Oh, huh. I hadn't known that was how news worked (though it makes sense)—I was only thinking of location filters in terms of picking a specific system from among them.

That doesn't seem too difficult a change.

quyykk commented on source/Timer.h:
```diff
@@ -0,0 +1,116 @@
+	void Step(PlayerInfo &player, UI *ui);
+
+private:
+	enum ResetCondition { NONE, PAUSE, LEAVE_ZONE, LEAVE_SYSTEM };
```
Enum class please

danaris replied:
Again, my lack of C++ experience shows: is it as simple as adding `class` after the `enum`? Or is more required here?

quyykk replied:
Yeah basically. You don't need to make the values all caps either

danaris replied:
...Is there a Style Guide on that last part? Because most, though not all, such `enum`s do currently seem to have all-caps names...

quyykk commented on source/PlayerInfo.cpp:
```diff
@@ -4142,6 +4143,15 @@ void PlayerInfo::StepMissions(UI *ui)
+	for(Mission &mission : missions)
+		for(Timer &timer : mission.Timers())
+			if(!timer.IsComplete())
+				timer.Step(*this, ui);
```
Step should be function to check whether the timer has completed.

danaris replied:
....So you'd want `Step()` to just have right at the start of it a short-circuit check for completion, in which case it returns a boolean? (Not sure whether `true` or `false` would make most sense...)

What, exactly, would that boolean value be for at that point? Would it be just as good simply to move the completion check into `Step()` and _not_ have it return a value?

quyykk replied:
> So you'd want Step() to just have right at the start of it a short-circuit check for completion

It already has one :) No need to return anything either.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,271 @@
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset")
+		{
+			if(child.Size() > 1 && child.Token(1) == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(child.Size() > 1 && child.Token(1) == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(child.Size() > 1 && child.Token(1) == "pause")
+				resetCondition = ResetCondition::PAUSE;
```
I think you can simplify this a bit. Are there any cases for timers to require say uncloaked and being in a zone but only resetting the timer if the ship leaves the zone? I would expect like most if not all timers to apply any reset if any of their conditions are no longer met.

danaris replied:
That would be `reset pause`—I can make that the default, and add an explicit `reset never` option instead, if you think that makes more sense.

I don't know what all the use cases would be; I believe that when I wrote this, there had been a Discord conversation a day or two previously that had given some ideas (such as the "uncloaked" part, and having to wait near a particular spöb), so I was just trying to make sure it would _at least_ accommodate those.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,271 @@
+	}
+	if(proximity > 0.)
+	{
+		Point center = Point(0, 0);
```
```suggestion
		Point center;
```

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,271 @@
+		}
+	}
+	isActive = true;
+	timeElapsed += (1. / 60.);
```
as mentioned, this should be counting frames instead of seconds

quyykk reviewed (commented):
Nice!

Timers basically have a condition that they must fulfill to run (in system, in zone, idling, etc.). If they stop fulfilling this, the timer should by default stop and reset IMO. I think that's a good default to have, because that's usually what you want.

danaris commented on source/Mission.h:
```diff
@@ -254,6 +257,8 @@ class Mission {
 	// NPCs:
 	std::list<NPC> npcs;
+	// Timers:
+	std::list<Timer> timers;
```
This is where my lack of C++ experience shows—I honestly don't understand the differences between the various STL collection types. I would gladly bow to superior experience in this.

danaris replied:
Changing to a `vector`.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,271 @@
+// so the time to wait will be saved fully calculated, and with any elapsed time subtracted
+void Timer::Save(DataWriter &out) const
+{
+	// If this NPC should no longer appear in-game, don't serialize it.
```
Ayep. Whoops!

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,271 @@
+	return result;
+}
+
+uint64_t Timer::TimeToWait() const
```
Reasonable. I think I wrote all the accessors up front, which has been my habit in other projects, where it's more usual to expect all of them to be needed at least _somewhere_.

danaris commented on source/Timer.h:
```diff
@@ -0,0 +1,116 @@
+*/
+
+#ifndef ES_TIMER_H_
+#define ES_TIMER_H_
```
...OK; I was just copying what I saw in some of the other files. Easy enough to change.

danaris commented on source/Timer.h:
```diff
@@ -0,0 +1,116 @@
+
+	// Used for holding the calculated time to wait and current timer value when it's actually active
+	int64_t timeToWait = 0;
+	double timeElapsed = 0.;
```
I had set it up to have the time specified in seconds in the file (which seemed the natural unit); it did occur to me later that that would make this unusual, as most if not all other timings are indeed specified in frames. 

So I guess that's the real question: Should the specification in the data file be in frames, too, for consistency's sake?

quyykk replied:
Yes please. Everything in frames

danaris replied:
Gotcha.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,253 @@
+	{
+		if(child.Token(0) == "idle")
+			requireIdle = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
```
Not sure but `"source"` sounds better imo? idk though

danaris replied:
Hm. "Source" seems to imply that it's coming _from_ the system where you have to wait. I would almost say "destination" fits better, but in both cases I think (as with the trigger, below) it is better to have it be something distinct from the specifiers for the mission itself, to reduce confusion. 

I suppose we could change it to something even more descriptive, like `wait in`? Though that has the disadvantage of requiring quoting...

On the whole, I'm inclined to stick with `system` unless you really think it should be changed.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,253 @@
+			requireIdle = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "systems")
```
Either way, this should use the same token as above.

danaris replied:
Gotcha. That seems reasonable.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,253 @@
+				resetCondition = ResetCondition::PAUSE;
+		}
+		// We keep "on timeup" as separate tokens so that it's compatible with MissionAction syntax
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
```
`on complete`?

danaris replied:
I chose `on timeup` so that it wouldn't cause confusion with the mission-level `on complete` trigger; I think that it both makes it a little easier for content creators to keep track of, and potentially allows for better future-proofing, as AIUI it gives the MissionAction object a distinct `trigger`.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,253 @@
+		Point center;
+		if(proximityCenter)
+		{
+			const StellarObject *proximityObject = system->FindStellar(proximityCenter);
```
A system might have multiple instances of the given planet, with the classic example being ringworlds. Presumably, it should apply to every single stellar object with the given planet, not just one.

danaris replied:
Ooh, I hadn't realized those were all instances of the same planet, though it makes sense. 

...So how does one go about getting _all_ the instances of a planet within a system? I'm not seeing a method on either System or Planet to do that. Is just iterating through `System::Objects()` the best way?

quyykk replied:
> Is just iterating through System::Objects() the best way?

yup

danaris replied:
Got this implemented, and tested it at Alta Hai—all looked good.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,253 @@
+		}
+	}
+	isActive = true;
+	timeElapsed += 1;
```
```suggestion
	++timeElapsed;
```

danaris replied:
I am curious about using the prefixed version—is there a specific performance or logic reason for doing it that way, rather than `timeElapsed++`? Or is it simply a style choice?

quyykk replied:
For performance reasons in the general case yes, for example with iterators it matters. For ints it doesn't matter, but for consistency preincrement is also used there.

quyykk commented on source/Timer.h:
```diff
@@ -0,0 +1,110 @@
+class Timer {
+public:
+	Timer() = default;
+	Timer(const DataNode &node, Mission *mission);
+	// Set up the timer from its data file node
+	void Load(const DataNode &node, Mission *mission);
```
You could take the mission by `const` here

danaris replied:
OK; that does make sense, as it's only there to pass to `GameAction::Do()`. Thanks.

danaris commented on source/MainPanel.cpp:
```diff
@@ -144,6 +144,9 @@ void MainPanel::Step()
 
 	engine.Step(isActive);
 
+	if(isActive)
+		player.StepMissionTimers(GetUI());
```
The `Timer::Step()` method, which this eventually calls, needs access to the UI object to pass to `MissionAction::Do()`; as a Panel, the MainPanel already has access to it (through the `GetUI()` method), while AFAICT, Engine does not.

If this call really does need to move inside `Engine::Step()`, what's the best way to go about getting it the UI?

quyykk replied:
oh whoops I thought engine had access, nvm sry my bad

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,263 @@
+	if(timeElapsed >= timeToWait)
+	{
+		action.Do(player, ui, mission);
+		player.Conditions().Add("timer: " + name + ": complete", 1);
```
Hm. As it stands, the syntax for the timer specification would become ambiguous if we allowed for a nameless timer—currently, it's `timer "timer name" base [rand]`.

I was more or less assuming that a timer would _have_ to have a name, though I can see that that's not really reflected in the Load() code.

Which do you think is a better course here—explicitly enforce that a timer must have a name (and if so, how?), or explicitly support nameless timers (and again, if so, how to deal with the ambiguity?)?

quyykk replied:
You could move the base/rand attributes as child of `timer` instead maybe?

danaris replied:
Hm. OK, yeah; I guess that will work fine, and doesn't seem too complicated. I will make that change.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,263 @@
+		return;
+
+	int64_t timeRemaining = timeToWait - timeElapsed;
+	out.Write("timer", name, timeRemaining);
```
I'm trying to come up with a situation where a player could save and reload, where the timer has _any_ reset condition, without the timer having already reset...? After all, once they land, they would trigger _all_ available reset conditions.

quyykk replied:
Yeah but isn't there the none reset condition that would never reset? Or did I misunderstand

danaris replied:
If it never resets, then you _have_ to save the time remaining—it shouldn't reset only if the player saves and reloads.

quyykk replied:
rip I'm dumb, that's what lack of sleep does :pensive:

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,263 @@
+				if(proximityObject.HasValidPlanet())
+					centers.push_back(proximityObject.Position());
+		}
+		for(Point &center : centers)
```
Ah, good call.

If I do that, is there still anything here that should be made `const &`?

danaris commented:
It occurred to me that it might also be useful, at least sometimes, to have the ability to fire an action when a timer resets (to let the player know that they're no longer on target or whatever), so I've added that.

quyykk commented on source/Mission.cpp:
```diff
@@ -856,6 +860,10 @@ bool Mission::IsSatisfied(const PlayerInfo &player) const
 		if(!npc.HasSucceeded(player.GetSystem()))
 			return false;
+	// All timers must be complete
+	for(const Timer &timer : timers)
+		if(!timer.IsComplete())
+			return false;
```
Do we want to require this?

danaris replied:
Initially I was thinking yes, but it does make sense to allow for "optional" timers, so I will add an "optional" attribute that can be added to circumvent this check. (Thus still defaulting to the timer being required to complete the mission.)

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+	if((system && player.Flagship()->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(player.Flagship()->GetSystem())))
+	{
+		bool didReset = ResetOn(Timer::ResetCondition::LEAVE_SYSTEM);
+		if(didReset && (repeatReset || !resetFired))
+		{
+			resetAction.Do(player, ui, mission);
+			resetFired = true;
+		}
+		isActive = false;
```
True—is it worth factoring this out into a separate convenience method?

quyykk replied:
can't it be added to ResetOn?

quyykk replied:
if not then yeah

danaris replied:
Hm. It looks like it probably can, yeah. Not sure why I didn't think to do that before; I was probably thinking that ResetOn should be more generic and modular.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+			for(const StellarObject &proximityObject : system->Objects())
+				if(proximityObject.HasValidPlanet())
+				{
+					double dist = (player.Flagship()->Position() - proximityObject.Position()).Length();
```
Ah—so it does. Thanks.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+	if(timeElapsed >= timeToWait)
+	{
+		action.Do(player, ui, mission);
+		if(!name.empty())
+			player.Conditions().Add("timer: " + name + ": complete", 1);
```
Sounds reasonable.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,307 @@
+
+	this->mission = mission;
+
+	timeToWait = base + Random::Int(rand);
```
Shouldn't this be done after loading the variables from the data file?

danaris replied:
Oh, whoops—this is left over from before I moved the setting of `base` into a child.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,307 @@
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity")
```
Hm. Are there other places that take a LocationFilter that _must_ be a Planet, not a System? I suppose either way, this could do that—it would just have to fail to load the Timer altogether if the LocationFilter doesn't resolve to one or more Planets.

quyykk replied:
> it would just have to fail to load the Timer altogether

Not necessarily, it can just do nothing, like in other places that allow a LocationFilter.

quyykk commented on source/Timer.cpp:
```diff
@@ -0,0 +1,320 @@
+{
+	if(isComplete)
+		return;
+	if((system && player.Flagship()->GetSystem() != system) ||
```
Check if there is a flagship first. This code will crash when the player dies.

quyykk reviewed (commented):
The Timer files needs to be added to the CodeBlocks project as well (EndlessSkyLib.cbp).

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,320 @@
+		timeElapsed = 0;
+		timeToWait = base + Random::Int(rand);
+		if(repeatReset || !resetFired)
+			{
+				resetAction.Do(player, ui, mission);
+				resetFired = true;
+			}
+			isActive = false;
```
Whoops. Got it.

danaris commented:
> The Timer files needs to be added to the CodeBlocks project as well (EndlessSkyLib.cbp).

OK; done.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+			if(child.Size() > 2)
+				rand = static_cast<uint32_t>(child.Value(2));
+
+			timeToWait = base + Random::Int(rand);
```
Need to check that the random value is >= 1.

danaris replied:
Sorry, for clarity: do you mean `rand` or the return value of `Random::Int(rand)`?

Amazinite replied:
Random::Int(0) throws an exception.

Amazinite replied:
The random value should also be stored and only added at instantiation time, not at load time. By doing it at load time, every instance of this timer will have the same random offset until the player relaunches the game.

danaris replied:
Gotcha. Will add a check for that.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+			}
+			if(child.HasChildren())
+			{
+				const DataNode &firstGrand = (*child.begin());
+				proximityCenter = GameData::Planets().Find(firstGrand.Token(0));
+				// If this is null, then it's not a planet by name, it's a LocationFiler
+				if(proximityCenter == nullptr)
+					proximityCenters.Load(firstGrand);
```
Not a fan of first checking if the first token is a planet and loading the first node as a location filter if the first token isn't recognized as a planet. Even if this was fine, this won't work, as a planet may not be loaded yet by the time this mission is loading, meaning we could falsely assume that the token we were given is for a planet.

The syntax should probably look more something like this:
```html
proximity [<planet name>]
	[<location filter>]
"proximity settings" <distance#> [close | far]
```

That way the handling of `proximity` can be similar to the handling of `source` or `destination` in Mission, or of `system` in this PR.

danaris replied:
Ah, I forgot that loading order is not guaranteed. I'll rework that.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
+			action.Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			resetAction.Load(child);
+
```
```suggestion
			resetAction.Load(child);
		else
			child.PrintTrace("Skipping unrecognized attribute:");
```

danaris replied:
(Adding this locally because I changed the previous line)

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+				out.Write("proximity", proximityCenter->Name());
+			else if(!proximityCenters.IsEmpty())
+			{
+				out.BeginChild();
+				{
+					proximityCenters.Save(out);
+				}
+				out.EndChild();
```
LocationFilter::Save already indents the child writing.
```suggestion
				out.Write("proximity");
				proximityCenters.Save(out);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+
+		action.Save(out);
+		resetAction.Save(out);
+
```
```suggestion
```

Amazinite replied:
This whitespace is still here.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+	}
+	if(requireIdle)
+	{
+		bool shipIdle = (!player.Flagship()->IsThrusting() && !player.Flagship()->IsSteering()
+						&& !player.Flagship()->IsReversing());
+		for(const Hardpoint &weapon : player.Flagship()->Weapons())
+			shipIdle &= !weapon.WasFiring();
```
This isn't exactly idle in the same sense that idle is currently used with the timer ship. This isn't checking the player's speed, meaning the player could be whizzing away from the system center but not taking any actions and the timer would still tick down. Check that flagship.Velocity().LengthSquared() is below some threshold^2 (squaring is cheaper than square rooting, which Point.Length() does), potentially defined by the idle keyword. The existence of the idle keyword should still set requireIdle, but an optional value of idle could define the velocity you're supposed to be below, defaulted to some low value if it isn't present.

danaris replied:
Makes sense. Just to check, would it be generally considered more desirable if "idle" also meant "not firing"...?

Amazinite replied:
I think it'd make sense to make that an option.

danaris replied:
OK. I've added a separate node for that.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+		}
+	}
+	isActive = true;
+	++timeElapsed;
+	if(timeElapsed >= timeToWait)
```
You could just increment in the comparison. The prefixed ++ increments and then compares (whereas if it were `var++ >= value` it would compare and then increment).
```suggestion
	if(++timeElapsed >= timeToWait)
```

Amazinite reviewed (changes requested):
Comments should be formatted as sentences. That is, include periods for statements.

danaris commented on source/Timer.h:
```diff
@@ -0,0 +1,121 @@
+	// Set to true when all the conditions are met
+	bool isActive = false;
+
+	const Mission *mission = nullptr;
```
OK, makes sense.

danaris commented on source/Timer.h:
```diff
@@ -0,0 +1,121 @@
+private:
+	// The name of the timer
+	std::string name;
+	// The basic amount of time to wait, with the optional random values
+	uint64_t base = 0;
+	uint32_t rand = 0;
```
The only drawback to this that I can see is that it eliminates the option to recalculate the random value on timer reset (it'll just reset to the initially-calculated random value). I don't think this is a showstopper, but it seems worth mentioning.

Amazinite replied:
It seems odd to me that a random value would be applied every time the timer resets. Typically, from looking at other locations where random values are used, random values are something that's applied once at instantiation. Now perhaps none of the other uses of random values are similar to this, but consider what adding the random value on reset actually means: if the random value is added whenever the timer resets, you could potentially decrease the remaining time to wait by resetting the timer and rolling a lower random value.

danaris commented on source/Timer.h:
```diff
@@ -0,0 +1,121 @@
+	bool resetFired = false;
+
+	// Used for holding the calculated time to wait and current timer value when it's actually active
+	int64_t timeToWait = 0;
+	int64_t timeElapsed = 0.;
```
Gotcha. This is an area where my C++-fu is weak.

danaris commented on source/Timer.h:
```diff
@@ -0,0 +1,121 @@
+	// The filter to use for which planet(s) to hang out near if a single planet isn't set
+	LocationFilter proximityCenters;
+
+	// The action to be performed when the timer elapses
+	MissionAction action;
+	// The action to be performed when the timer resets
+	MissionAction resetAction;
```
Hm. Yes, I think that makes sense—particularly since I've already thought that we _might_ want an "on timer begin" action (but haven't added it because it's easy to simulate by adding a second timer with the same parameters but a much shorter time).

Is it better in a case like this to overload the `Trigger` enum name (which I _think_ I understand should still be valid, since it would be within the Timer class...?), or to make a new name like `TimerTrigger` for full clarity?

Amazinite replied:
> Is it better in a case like this to overload the Trigger enum name, or to make a new name like TimerTrigger for full clarity?

Honestly, either way is fine. We already have Trigger defined in both NPC and Mission, but it's not like it hurts to be specific.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+	++timeElapsed;
+	if(timeElapsed >= timeToWait)
+	{
+		if(!name.empty())
+			player.Conditions().Add("timer: " + name + ": complete", 1);
```
At this point, I don't honestly remember why I thought that would be important. I'll remove it.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,322 @@
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity")
+		{
+			// If the proximity node specifies a single planet, we take that;
+			// otherwise, we look to its child for a LocationFilter
+			if(child.Size() > 1)
+				proximityCenter = GameData::Planets().Find(child.Token(1));
+			else if(child.HasChildren())
+			{
+				const DataNode &firstGrand = (*child.begin());
+				proximityCenters.Load(firstGrand);
+			}
+		}
```
Hm. I was getting the child because I assumed that passing the entire node, with its token being `proximity`, would cause LocationFilter to error. If that's how it's _supposed_ to be called, that's certainly more convenient!

Amazinite replied:
When calling Load on any data object, they generally assume that the data node you pass it is correct and that all children nodes should be parsed.

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+	Timer Instantiate(std::map<std::string, std::string> &subs,
+					const System *origin, int jumps, int64_t payload) const;
+	// Progress the timer within the main loop.
+	void Step(PlayerInfo &player, UI *ui, Mission *mission);
```
I'd say just pass the mission as a reference. I know that MissionAction accepts mission as a pointer, but that's because GameAction (a member of MissionAction) can potentially be given a nullptr for mission in other locations. In the case of Timer, we know we're always getting something. A Timer can't exist except as a part of a mission, after all.

Could make it const too since Timer only needs the mission so it can pass it off to MissionAction, and MissionAction takes a `const Mission *`.
```suggestion
	void Step(PlayerInfo &player, UI *ui, const Mission &mission);
```

Amazinite commented on source/PlayerInfo.cpp:
```diff
@@ -4140,6 +4141,14 @@ void PlayerInfo::StepMissions(UI *ui)
+{
+	for(Mission &mission : missions)
+		for(Timer &timer : mission.Timers())
+			timer.Step(*this, ui, &mission);
```
```suggestion
			timer.Step(*this, ui, mission);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+
+
+
+void Timer::Step(PlayerInfo &player, UI *ui, Mission *mission)
```
```suggestion
void Timer::Step(PlayerInfo &player, UI *ui, const Mission &mission)
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+
+
+
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, Mission *mission)
```
```suggestion
void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, const Mission &mission)
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+		timeElapsed = 0;
+		if(repeatReset || !resetFired)
+		{
+			actions[RESET].Do(player, ui, mission);
```
```suggestion
			actions[RESET].Do(player, ui, &mission);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+	isActive = true;
+	if(++timeElapsed >= timeToWait)
+	{
+		actions[TIMEUP].Do(player, ui, mission);
```
There may not be a TIMEUP action defined.
```suggestion
		auto it = actions.find(TIMEUP)
		if(it != actions.end())
			it->second.Do(player, ui, &mission);
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+		LEAVE_ZONE,
+		LEAVE_SYSTEM
+	};
+	void ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, Mission *mission);
```
```suggestion
	void ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, const Mission &mission);
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+
+
+private:
+	// The default speed for determining whether the player is idle
+	const double DEFAULT_MAX_SPEED = 25.;
```
There's really no need for this to be a constant...
```suggestion
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+	// Whether the timer requires the player to be idle.
+	bool requireIdle = false;
+	// The speed threshold the player's flagship must be under to count as "idle".
+	double idleMaxSpeed = DEFAULT_MAX_SPEED;
```
... just give idleMaxSpeed a default value of 25.
```suggestion
	double idleMaxSpeed = 25.;
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+class Timer {
+public:
+	Timer() = default;
+	Timer(const DataNode &node, const Mission *mission);
+	// Set up the timer from its data file node.
+	void Load(const DataNode &node, const Mission *mission);
```
No longer used.
```suggestion
	Timer(const DataNode &node);
	// Set up the timer from its data file node.
	void Load(const DataNode &node);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+
+using namespace std;
+
+Timer::Timer(const DataNode &node, const Mission *mission)
+{
+	Load(node, mission);
+}
+
+
+
+void Timer::Load(const DataNode &node, const Mission *mission)
```
```suggestion
Timer::Timer(const DataNode &node)
{
	Load(node);
}



void Timer::Load(const DataNode &node)
```

Amazinite commented on source/Mission.cpp:
```diff
@@ -288,6 +288,8 @@ void Mission::Load(const DataNode &node)
 		else if(child.Token(0) == "npc")
 			npcs.emplace_back(child);
+		else if(child.Token(0) == "timer")
+			timers.emplace_back(child, this);
```
```suggestion
			timers.emplace_back(child);
```

Amazinite commented on source/Mission.cpp:
```diff
@@ -856,6 +860,10 @@ bool Mission::IsSatisfied(const PlayerInfo &player) const
 	for(const NPC &npc : npcs)
 		if(!npc.HasSucceeded(player.GetSystem()))
 			return false;
+	// All timers must be complete
```
```suggestion
	// All non-optional timers must be complete.
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+{
+	if(isComplete)
+		return;
+	Ship *flagship = player.Flagship();
```
```suggestion
	const Ship *flagship = player.Flagship();
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+	// The name of the timer.
+	std::string name;
+	// The basic amount of time to wait, with the optional random values.
+	uint32_t timeToWait = 0;
+	uint32_t rand = 0;
```
I said normal. :)
```suggestion
	int timeToWait = 0;
	int rand = 0;
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+	bool resetFired = false;
+
+	// Used for holding the current timer value when it's actually active.
+	uint32_t timeElapsed = 0.;
```
```suggestion
	int timeElapsed = 0.;
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			timeToWait = static_cast<int64_t>(child.Value(1));
+			if(child.Size() > 2)
+				rand = static_cast<uint32_t>(child.Value(2));
```
```suggestion
			timeToWait = child.Value(1);
			if(child.Size() > 2)
				rand = child.Value(2);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+		}
+		if(requireIdle)
+		{
+			if(idleMaxSpeed == DEFAULT_MAX_SPEED)
+				out.Write("idle");
+			else
+				out.Write("idle", sqrt(idleMaxSpeed));
```
No harm in just outputting the default.
```suggestion
			out.Write("idle", sqrt(idleMaxSpeed));
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			if(proximityCenter != nullptr)
```
Pointers implicitly convert to bools in if statements.
```suggestion
			if(proximityCenter)
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+
+void Timer::Load(const DataNode &node, const Mission *mission)
+{
+	if(node.Size() > 1)
+		name = node.Token(1);
+
+
```
The name isn't used anymore.
```suggestion
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+		return;
+
+	int64_t timeRemaining = timeToWait - timeElapsed;
+	if(!name.empty())
+		out.Write("timer", name);
+	else
+		out.Write("timer");
```
```suggestion
	out.Write("timer");
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,326 @@
+						const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.name = name;
```
```suggestion
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+private:
+	// The default speed for determining whether the player is idle
+	const double DEFAULT_MAX_SPEED = 25.;
+	// The name of the timer.
+	std::string name;
```
```suggestion
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+	LocationFilter proximityCenters;
+
+	// The possible triggers for actions on this timer.
+	enum TimerTrigger {TIMEUP, RESET};
```
Define this as an enum class like how NPC defines it.

https://github.com/endless-sky/endless-sky/blob/6afcba37f13f9c37f2c0411f3ef0f908cff7f12f/source/NPC.h#L53-L65

danaris replied:
As opposed to basic enums like how Mission defines them ^_^;; Gotcha. Change made.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+	result.requirePeaceful = requirePeaceful;
+
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
```
These are always the same for non-instantiated vs newly instantiated timers. The only things that need instantiated for missions are those things that can be changed by the data definition.
```suggestion
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
```
MissionActions need to be validated during instantiation. (Copying code from Mission::Instantiate.)
```suggestion
	auto ait = actions.begin();
	for( ; ait != actions.end(); ++ait)
	{
		reason = ait->second.Validate();
		if(!reason.empty())
			break;
	}
	if(ait != actions.end())
	{
		Logger::LogError("Instantiation Error: Timer action \"" + TriggerToText(ait->first) + "\" in mission \""
			+ Identifier() + "\" uses invalid " + std::move(reason));
		return result;
	}
	for(const auto &it : actions)
		result.actions[it.first] = it.second.Instantiate(subs, sourceSystem, jumps, payload);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+			actions[TimerTrigger::RESET].Load(child);
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+
```
Ditto.
```suggestion
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+#include "UI.h"
+
+using namespace std;
+
```
```suggestion



```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,327 @@
+			systems.Save(out);
+		}
+		if(requireIdle)
+		{
+			out.Write("idle", sqrt(idleMaxSpeed));
+		}
```
```suggestion
			out.Write("idle", sqrt(idleMaxSpeed));
```

RisingLeaf commented on data/sheragi/archaeology missions.txt:
```diff
@@ -339,17 +339,19 @@ mission "Sad Archie"
+				scene "scene/eso0"
+				`You dream that you leave this star system and return to Zug, but when you land there everything is different: the cities you remember are gone, and in their place are other cities with strange architecture, populated by gargantuan dragon-like creatures.`
+				`	Time speeds up, and you watch as the dragon cities grow and spread out. You see one city buried under a massive lava flow, starships buzzing around it like a cloud of flies to carry people and goods to safety. Other cities are destroyed by war, and then abruptly the dragon-people are all gone and their cities lie vacant. Buildings crumble, and the forests reclaim the land.`
+				`	Then you wake up and find yourself back in your ship, in the Ildaria system. Weird.`
```
Data changes with code :raised_eyebrow:

danaris replied:
Is this Not Done™? I can break this out into a separate PR if that's preferable. It's here for two reasons—first, to demonstrate the usage, and second, to make sure that _as soon as_ this feature is merged, Sad Archie can make use of it and stop being a crapshoot.

RisingLeaf replied:
Well I dont like it, there might be different opinions though

Hurleveur replied:
it's litteraly the main use case for this, makes sense we'd use it

RisingLeaf commented on source/PlayerInfo.cpp:
```diff
@@ -44,6 +44,7 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include "StartConditions.h"
 #include "StellarObject.h"
 #include "System.h"
+#include "Timer.h"
```
not really needed

RisingLeaf commented on source/PlayerInfo.cpp:
```diff
@@ -4165,6 +4166,15 @@ void PlayerInfo::StepMissions(UI *ui)
+void PlayerInfo::StepMissionTimers(UI *ui)
+{
+	for(Mission &mission : missions)
+		for(Timer &timer : mission.Timers())
+			timer.Step(*this, ui, mission);
```
if we would move this into mission, player would not have to touch the timer class

RisingLeaf replied:
So just to explain this, I think if possible double sided includes should be avoided, in this case player uses timer and timer uses player, suboptimal imo

danaris replied:
I fully approve of the principle; however, I'm still having a little trouble implementing this—specifically, in `MainPanel.cpp:106` where `PlayerInfo::StepMissionTimers()` is called, I would need to replace it with iterating over the Missions in the `player` object, then calling a new `StepTimers()` method on each one. So far so good, but `PlayerInfo::Missions()` returns a list of `const Mission &` objects, and `StepTimers()` is a mutating method, so if I understand correctly, it cannot be `const`, and thus cannot be called on a `const Mission &`.

RisingLeaf replied:
That's not what I meant, keep this function but loop through the timers inside the mission class

RisingLeaf commented on source/Timer.cpp:
```diff
@@ -0,0 +1,346 @@
+using namespace std;
+
+namespace {
+	string TriggerToText(Timer::TimerTrigger trigger)
+	{
+		switch(trigger)
+		{
+			case Timer::TimerTrigger::TIMEUP:
+				return "on timeup";
+			case Timer::TimerTrigger::RESET:
+				return "on reset";
+			default:
+				return "unknown trigger";
+		}
+	}
```
you can make this a const std::map

danaris replied:
This is an exact mirror of the way it's done in both NPC and Mission—should both of those be refactored to be `const std::map`s as well?

RisingLeaf replied:
Now I am unsure

Amazinite replied:
Let's leave this as is for now. We can refactor in the future if we want.

RisingLeaf commented on source/Timer.cpp:
```diff
@@ -0,0 +1,346 @@
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset" && child.Size() > 1)
+		{
+			const string &value = child.Token(1);
+			if(value == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(value == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(value == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(value == "pause")
+				resetCondition = ResetCondition::PAUSE;
+			else
+				child.PrintTrace("Skipping unrecognized attribute:");
```
wouldnt it be possible to have multiple of these possible?

danaris replied:
Hhhhypothetically, I believe so? I think the only one that could coexist with others in practice is `PAUSE`—leaving the system and leaving the zone are basically just different degrees of moving too far away, but pausing the count (ie, moving if it requires idle, firing if it requires peaceful) _could_ be tested for separately.

Does that seem like it's worth refactoring to add the functionality?

RisingLeaf replied:
Nah

RisingLeaf commented on source/Timer.cpp:
```diff
@@ -0,0 +1,346 @@
+				if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+					proximityCenters.Matches(proximityObject.GetPlanet())))
+				{
+					double dist = flagship->Position().Distance(proximityObject.Position());
```
I think we should cache the proximityobject

danaris replied:
Do you mean we should build a cache of all the specified proximityObjects?

I think that should be possible, but once again I'm having trouble with type modifiers—if I add a `std::list<StellarObject> proximityCache` to the Timer definition, and attempt to populate it in the `Instantiate()` method by simply copying the above loop in, I can't seem to get it to allow me to call `proximityCache.push_back(proximityObject)`, even if I try changing the list to hold `const`s...

RisingLeaf replied:
Did you try a pointer?

danaris replied:
Which kind of pointer? Try it how? I did try to figure out if it could be done with a `shared_ptr` (with little success), but I'm still pretty new at handling all the different kinds of pointers and references that modern C++ supports. (I'd also be happy to discuss it further on Discord when & if that's convenient.)

RisingLeaf replied:
I was thinking about a const StellarObject*

RisingLeaf replied:
Also a std::set might be the right thing for the job?

danaris replied:
With the cache declared as `std::list<const StellarObject *> proximityCache;`, and the loop as

```
	for(const StellarObject &proximityObject : system->Objects())
		proximityCache.push_back(proximityObject);
```
(leaving out the lengthy conditional, which is presumably irrelevant for this problem), I get this build error

```
endless-sky/source/Timer.cpp:242:19: error: no matching member function for call to 'push_back'
                        proximityCache.push_back(proximityObject);
                        ~~~~~~~~~~~~~~~^~~~~~~~~
/usr/include/c++/v1/list:986:10: note: candidate function not viable: 'this' argument has type 'const std::list<const StellarObject *>', but method is not marked const
    void push_back(value_type&& __x);
         ^
/usr/include/c++/v1/list:1011:10: note: candidate function not viable: 'this' argument has type 'const std::list<const StellarObject *>', but method is not marked const
    void push_back(const value_type& __x);
         ^
1 error generated.
```

(cleaned up slightly for readability, but not removing any of the salient details)

RisingLeaf replied:
You are trying to write data in a const marked function?

danaris replied:
Ah! You're right; thanks for catching that dumb mistake. Correcting it to `result.proximityCache.push_back(&proximityObject)` appears (provisionally) to work.

I will see about adjusting the rest of the code to use that cache.

RisingLeaf reviewed (commented):
Definitely well commented _in the header_ could need some comments in the implementation file

danaris commented:
I've added some comments to `Timer.cpp`. Please let me know if they seem helpful or if they still need work.

RisingLeaf commented:
> I've added some comments to `Timer.cpp`. Please let me know if they seem helpful or if they still need work.

Better 👍

RisingLeaf commented on source/Timer.h:
```diff
@@ -0,0 +1,133 @@
+	bool isActive = false;
+
+	// A place to cache the proximity object(s) so we don't have to test for them every time
+	std::list<const StellarObject *> proximityCache;
```
This should be a std::set instead as we don't need duplicate objects and don't need to search objects (at least that's what I recall 😅 )

danaris replied:
Gotcha. I think the latest commit should do that correctly.

RisingLeaf commented on source/Mission.cpp:
```diff
@@ -1132,6 +1140,20 @@ void Mission::UpdateNPCs(const PlayerInfo &player)
+	return timers;
+}
+
+// Iterate through the timers and progress them if applicable
```
```suggestion
// Iterate through the timers and progress them if applicable.
```

RisingLeaf commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,48 @@
+		call "Depart"
+		call "Land"
+		call "Depart"
+		watchdog 10000
```
not supposed to be here

danaris replied:
Just the `watchdog` line?

RisingLeaf replied:
yes

RisingLeaf commented on source/Mission.h:
```diff
@@ -160,6 +161,10 @@ class Mission {
 	// takes off from a planet, they should be added to the active ships.
 	const std::list<NPC> &NPCs() const;
+	// Get a list of timers associated with this mission.
+	std::list<Timer> &Timers();
```
again I dont think we need a list here, a std set would be enough?
also we dont need this getter anymore right?

RisingLeaf replied:
> And am I correct in my understanding that the (primary?) difference between the two is that list can have duplicate members, while set cannot?
yes

> Again, I'm afraid I'm not well-acquainted with the distinctions between various STL collection types—I simply copied the usage for NPCs. Is it that there's something different about how Timers work than NPCs that means they should have a set while NPCs have a list, or would this also apply to the Mission's NPC collection?

dont know

RisingLeaf commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+		apply
+			"test: steps to wait" = 10 * 60
+		label "floating"
+		# empty input to make 10 seconds and one frame pass
```
this comment seems wrong?
also maybe give it a bit more time, just to be sure?

danaris replied:
Any idea how much? I'm afraid I've got no experience with the tests (these were actually written by someone else—I forget who, offhand).

danaris replied:
Boosted it to 30s.

Amazinite commented on source/Mission.h:
```diff
@@ -160,6 +161,8 @@ class Mission {
 	// Get a list of NPCs associated with this mission. Every time the player
 	// takes off from a planet, they should be added to the active ships.
 	const std::list<NPC> &NPCs() const;
+	// Iterate through the timers and progress them if applicable
```
```suggestion
	// Iterate through the timers and progress them if applicable.
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,133 @@
+
+
+
+// Class representing a timer for triggering mission actions
+// This requires a specification of how long to wait (base + randint(rand)),
+// a system to wait in, and whether to reset the timer upon leaving the system
+// and can optionally require the player to be idle (no acceleration), and/or
+// within or beyond a certain proximity to the system center or a
+// specific (named) system object.
```
```suggestion
// Class representing a timer for triggering mission actions. Timers count
// down a certain number of frames before triggering if the player meets the
// conditions for the timer starting (e.g. the player is moving slowly or is near
// a certain object), but may be reset by various actions the player takes. 
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,133 @@
+	// Set to true when all the conditions are met.
+	bool isActive = false;
+
+	// A place to cache the proximity object(s) so we don't have to test for them every time
```
```suggestion
	// A place to cache the proximity object(s) so we don't have to test for them every time.
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,133 @@
+
+	// Get whether the timer is optional to complete.
+	bool IsOptional() const;
+	// Get whether the timer is currently active.
```
```suggestion
	// Get whether the timer has completed.
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,133 @@
+	// The filter for the systems it can be for.
+	LocationFilter systems;
+
+	// If set, the mission the timer is attached to can be completed even without it.
```
```suggestion
	// If set, the timer is not a necessary objection for the completion of its mission.
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,133 @@
+	int timeElapsed = 0.;
+	// Set to true once the timer has run to completion so we don't keep trying to save or run it.
+	bool isComplete = false;
+	// Set to true when all the conditions are met.
```
```suggestion
	// Set to true when all the conditions are met for the timer to count down.
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,376 @@
+			requireIdle = true;
+			// We square the max speed value here, so it can be conveniently
+			// compared to the flagship's squared velocity length below.
+			if(child.Size() > 1)
+				idleMaxSpeed = child.Value(1) * child.Value(1);
```
```suggestion
			if(child.Size() > 1)
			{
				idleMaxSpeed = child.Value(1);
				idleMaxSpeed *= idleMaxSpeed;
			}
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,376 @@
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity" && child.Size() > 1)
+		{
+			proximityCenter = GameData::Planets().Get(child.Token(1));
+		}
+		else if(child.Token(0) == "proximity")
+		{
+			proximityCenters.Load(child);
+		}
```
```suggestion
		else if(child.Token(0) == "proximity" && child.Size() > 1)
			proximityCenter = GameData::Planets().Get(child.Token(1));
		else if(child.Token(0) == "proximity")
			proximityCenters.Load(child);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,376 @@
+	if(isComplete)
+		return;
+
+	int64_t timeRemaining = waitTime - timeElapsed;
+	out.Write("timer");
+	out.BeginChild();
+	{
+		out.Write("time", timeRemaining);
```
We need to separately save the time elapsed and the wait time instead of calculating a time remaining and saving that, as otherwise when we save the game we have no idea what to reset the timer to if it need reset, or it'll reset to the wrong value.

danaris replied:
OK. Adding this.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,376 @@
+
+
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+bool Timer::IsOptional() const
+{
+	return optional;
+}
```
These don't match their order in the header and are missing their comments.

danaris replied:
Got it. Order swapped & comments added.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,376 @@
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	// Check whether this timer actually wants to reset on the given condition.
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
```
This seem difficult to read to me. So we reset if the input condition matches the reset condition for this timer, but the input condition and reset condition can differ and we still reset the timer? I feel like calling ResetOn should just always reset the timer, and we only call the function when we need to, instead of calling the function at every possible chance and having the function figure out if it should actually reset.

danaris replied:
I believe I moved this in here to avoid duplication, as it would need to be checked every time a reset might happen.

I've added some comments to clarify the intention here, which I have pasted below:
```
// If the reset condition being passed in is LEAVE_ZONE (ie, the player is no longer in proximity),
// then we also reset if the timer's reset condition is PAUSE (ie, player no longer meets the criteria).
...
// If the reset condition being passed in is LEAVE_SYSTEM, then we also reset if either of the other two
// conditions is specified for the timer, since we are then necessarily outside the zone, and thus the
// clock is stopped.
```

Please let me know if you still think these tests should be broken back out to the call sites.

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,376 @@
+
+	// We also build a cache of the matching proximity object(s) for the instantiated Timer.
+	// This avoids having to do all these comparisons every Step().
+	for(const StellarObject &proximityObject : system->Objects())
+		if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+			proximityCenters.Matches(proximityObject.GetPlanet())))
+			result.proximityCache.insert(&proximityObject);
```
Nullptr exception if there is no system. Also, we only need to build a cache if we actually have a proximity center defined.
```suggestion
	if(system && (proximityCenter || !proximityCenters.isEmpty()))
		for(const StellarObject &proximityObject : system->Objects())
			if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
				proximityCenters.Matches(proximityObject.GetPlanet())))
				result.proximityCache.insert(&proximityObject);
```

TomGoodIdea commented on source/Timer.cpp:
```diff
@@ -0,0 +1,389 @@
+			requirePeaceful = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity" && child.Size() > 1)
+			proximityCenter = GameData::Planets().Get(child.Token(1));
+		else if(child.Token(0) == "proximity")
+			proximityCenters.Load(child);
```
save some checks
```suggestion
		else if(child.Token(0) == "system")
		{
			if(child.Size() > 1)
				system = GameData::Systems().Get(child.Token(1));
			else
				systems.Load(child);
		}
		else if(child.Token(0) == "proximity")
		{
			if(child.Size() > 1)
				proximityCenter = GameData::Planets().Get(child.Token(1));
			else
				proximityCenters.Load(child);
		}
```

TomGoodIdea commented on source/Timer.cpp:
```diff
@@ -0,0 +1,389 @@
+	if(system && (proximityCenter || !proximityCenters.IsEmpty()))
+		for(const StellarObject &proximityObject : system->Objects())
+			if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+				proximityCenters.Matches(proximityObject.GetPlanet())))
```
```suggestion
					proximityCenters.Matches(proximityObject.GetPlanet())))
```
looks clearer

TomGoodIdea commented on source/Timer.cpp:
```diff
@@ -0,0 +1,389 @@
+		bool shipIdle = true;
+		if(requireIdle)
+			shipIdle = (!flagship->IsThrusting() && !flagship->IsSteering()
+						&& !flagship->IsReversing() && flagship->Velocity().LengthSquared() < idleMaxSpeed);
```
```suggestion
				&& !flagship->IsReversing() && flagship->Velocity().LengthSquared() < idleMaxSpeed);
```

TomGoodIdea commented on source/Timer.cpp:
```diff
@@ -0,0 +1,389 @@
+	if(proximity > 0.)
+	{
+		bool inProximity = false;
+		if(proximityCache.size() > 0)
+			for(const StellarObject *proximityObject : proximityCache)
+			{
+				double dist = flagship->Position().Distance(proximityObject->Position());
+				if((closeTo && dist <= proximity) || (!closeTo && dist >= proximity))
+				{
+					inProximity = true;
+					break;
+				}
+			}
```
if proximityCache is empty, the loop doesn't do anything anyway; I'd also use a ternary for the proximity conditions
```suggestion
		for(const StellarObject *proximityObject : proximityCache)
		{
			double dist = flagship->Position().Distance(proximityObject->Position());
			if(closeTo ? dist <= proximity : dist >= proximity)
			{
				inProximity = true;
				break;
			}
		}
```

RisingLeaf replied:
If you really want to simplify the conditional just do `closeto != dist >= proximity`
Even though that would ignore the single case of == but that's not important for double values

danaris replied:
> if proximityCache is empty, the loop doesn't do anything anyway; I'd also use a ternary for the proximity conditions

While this is more concise code, it's also much harder to parse on first look, IMO. I think it would be much clearer to just assign the value of the ternary to `inProximity`, then test `inProximity` and break if true.

danaris commented:
Due to #10417, I cannot currently build & test ES on my own. Under the circumstances, I would have absolutely no objection to someone else picking up this PR and carrying it across the finish line; indeed, I would encourage it, and if there's anything I need to do to enable or facilitate that, please let me know and I will do my best.

tibetiroka commented:
There is an unaddressed review above.

danaris commented:
> Comments should be formatted as sentences. That is, include periods for statements.

I believe this should be completed.

> There is an unaddressed review above.

If there's still anything outstanding that genuinely needs resolving, can you point it out directly? There's been a lot of activity on this, and a few things are technically "unresolved", but AFAICT, just where there were differences of opinion on what's ideal, nothing that should actually be a blocker at this point.

tibetiroka commented:
It's failing the checks.

danaris commented:
> It's failing the checks.

It appears that the failing checks are for the integration tests that it adds; I'm afraid I'm not very familiar with the testing infrastructure, and Hurleveur was kind enough to write the initial test for me. I'd be happy to fix it with some guidance, since it does seem like a nice thing to have. Otherwise I would just remove the test entirely, which should make it pass the checks.

tibetiroka commented:
Are you sure the code works otherwise?

danaris commented:
> Are you sure the code works otherwise?

As noted in the PR, I have tested it in quite a variety of ways, and they all work.

ziproot commented:
The problem is it is testing a ten second timer for 30 seconds, which also happens to be the timeout, so it doesn’t have time to finish. I would assume 20 seconds should also work.

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+		call "Land"
+		call "Depart"
+		apply
+			"test: steps to wait" = 30 * 60
```
```suggestion
			"test: steps to wait" = 20 * 60
```

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+		apply
+			"test: steps to wait" = 30 * 60
+		label "floating"
+		# An empty input to make 30 seconds pass.
```
```suggestion
		# An empty input to make 20 seconds pass.
```

danaris commented:
Accepted those changes—let's see if the checks complete this time.

danaris commented:
> It's failing the checks.

Looks like ziproot's changes fixed this—though for some reason the Mac Intel CI check doesn't seem to have picked them up? Its copy of the integration test is still timing out at 30s, while the others are completing at around 20s.

And the Steam CI check is failing, with error messages that indicate it's definitely something to do with its configuration, and not my code.

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+		mission "10 second timer"
+			landing
+			timer "Why do I have a name?"
+				time 600
```
```suggestion
				time 60
```

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+test-data "Timer Test Mission"
+	category mission
+	contents
+		mission "10 second timer"
```
```suggestion
		mission "1 second timer"
```

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+					set "test: timer condition"
+
+
+test "Timer - Simple 10 seconds timer"
```
```suggestion
test "Timer - Simple 1 second timer"
```

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+		call "Land"
+		call "Depart"
+		apply
+			"test: steps to wait" = 20 * 60
```
```suggestion
			"test: steps to wait" = 2 * 60
```

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+		apply
+			"test: steps to wait" = 20 * 60
+		label "floating"
+		# An empty input to make 20 seconds pass.
```
```suggestion
		# An empty input to make two seconds pass.
```

ziproot commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,47 @@
+
+test "Timer - Simple 10 seconds timer"
+	status active
+	description "Tests if a timer will trigger a mission event in 10 seconds as planned."
```
```suggestion
	description "Tests if a timer will trigger a mission event in one second as planned."
```

ziproot reviewed (commented):
This cuts the timer down to one second, which should still test the feature while being much less likely to timeout. If this still times out, it’s likely a different issue.

danaris commented:
Excellent, that last one got the checks to pass.

TomGoodIdea commented on source/Timer.cpp:
```diff
@@ -0,0 +1,394 @@
+		else if(child.Token(0) == "idle")
+		{
+			requireIdle = true;
+			// We square the max speed value here, so it can be conveniently
+			// compared to the flagship's squared velocity length below.
+			if(child.Size() > 1)
+			{
+				idleMaxSpeed = child.Value(1);
+				idleMaxSpeed *= idleMaxSpeed;
```
```suggestion
			if(child.Size() > 1)
			{
				idleMaxSpeed = child.Value(1);
				// We square the max speed value here, so it can be conveniently
				// compared to the flagship's squared velocity length below.
				idleMaxSpeed *= idleMaxSpeed;
```

TomGoodIdea commented on source/Timer.cpp:
```diff
@@ -0,0 +1,394 @@
+		else if(child.Token(0) == "proximity settings" && child.Size() > 1)
+		{
+			proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
```
closeTo defaults to true, so no need to check for "close" here.
```suggestion
			if(child.Size() > 2 && child.Token(2) == "far")
				closeTo = false;
```

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+// a certain object), but may be reset by various actions the player takes.
+class Timer {
+public:
+	enum class TimerTrigger {
+		// The possible triggers for actions on this timer.
```
```suggestion
	// The possible triggers for actions on this timer.
	enum class TimerTrigger {
```

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	bool optional = false;
+	// Whether the timer requires the player to be idle.
+	bool requireIdle = false;
+	// The speed threshold the player's flagship must be under to count as "idle".
```
```suggestion
	// The square of the speed threshold the player's flagship must be under to count as "idle".
```

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	double idleMaxSpeed = 25.;
+	// Whether the timer requires the player to not be firing.
+	bool requirePeaceful = false;
+	// Whether the timer requires the player to be uncloaked to advance.
```
```suggestion
	// Whether the timer requires the player to be uncloaked.
```
This "to advance" makes it seem like this requirement works somewhat differently than the others.

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	bool requireUncloaked = false;
+	// What circumstances will reset the timer: leaving the system,
+	// leaving the proximity zone (if applicable), or any circumstance that stops the timer.
+	ResetCondition resetCondition = Timer::ResetCondition::PAUSE;
```
```suggestion
	ResetCondition resetCondition = ResetCondition::PAUSE;
```

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	bool resetFired = false;
+
+	// Used for holding the current timer value when it's actually active.
+	int timeElapsed = 0.;
```
```suggestion
	int timeElapsed = 0;
```

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	// Set to true once the timer has run to completion so we don't keep trying to save or run it.
+	bool isComplete = false;
+	// Set to true when all the conditions are met for the timer to count down.
+	bool isActive = false;
```
Shouldn't this default to true?
```suggestion
	bool isActive = true;
```

danaris replied:
No, I don't believe so. The timer doesn't become active until its conditions are met (right system, right proximity, etc).

TomGoodIdea replied:
hmm ok

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	// A place to cache the proximity object(s) so we don't have to test for them every time.
+	std::set<const StellarObject *> proximityCache;
+};
+
+
+
+#endif
```
```suggestion

```

TomGoodIdea commented on source/Timer.h:
```diff
@@ -0,0 +1,131 @@
+	bool closeTo = true;
+	// This specifies the radius around the proximity center to be checked.
+	double proximity = 0.;
+	// This specifies the system object to be the center; if this is nullptr and the filter is empty,
+	// but proximity is positive, then the system center is used instead.
```
I don't see any actual code for this.

danaris replied:
Hm, whoops. I guess it must have gotten lost when I added the code for proximity objects.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,394 @@
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			if(proximityCenter)
+				out.Write("proximity", proximityCenter->TrueName());
+			else if(!proximityCenters.IsEmpty())
+			{
+				out.Write("proximity");
+				proximityCenters.Save(out);
+			}
```
Interesting. Yes, I think that would work here.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,394 @@
+	int jumps, int64_t payload) const
+{
+	Timer result;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.proximityCenters = proximityCenters;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+	result.requirePeaceful = requirePeaceful;
```
requirePeaceful: Will do.

resetFired: I believe we need it? If we don't instantiate it, it can't be saved and restored, which means that a timer with a reset action that is supposed to fire only once can have that action fired as often as you want if you land between reset and timeup.

proximityCenter(s): I think you're right; we should be able to just save the cached planets.

TomGoodIdea replied:
> resetFired: I believe we need it? If we don't instantiate it, it can't be saved and restored

Why? If we don't instantiate it, it would just default to false. Is the base timer able to expire before the instantiation? If yes, should this value be carried to the instances? Or am I misunderstanding what resetFired is for?

danaris replied:
`resetFired` is for tracking whether this timer has ever reset itself—that is, it's started, and then stopped before the time ran out. That way, it can have an action that fires on reset, but no more than once.

TomGoodIdea replied:
OK, but the base timers others are instantiated from aren't really functional timers, are they?

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,394 @@
+
+	// We also build a cache of the matching proximity object(s) for the instantiated timer.
+	// This avoids having to do all these comparisons every Step().
+	if(system && (proximityCenter || !proximityCenters.IsEmpty()))
```
Ah....I think it wouldn't. I think this needs a second check for having multiple `System`s filtered, in which case all the matching `Planet`s in those systems should be cached.

Amazinite replied:
A system location filter could match a LOT of planets, so I don't think it makes sense to cache all possible matches. 
I think it's also worth noting that the method of using a location filter for named planets for the proximity check means we can't require that the player be in proximity to any particular object in the system; we're forcing it to be the case that each object also has a planet associated with it, which might not be the case for things like wanting the player to be close to a black hole or some other unique object in the system.

danaris commented on source/Timer.cpp:
```diff
@@ -0,0 +1,394 @@
+			shipIdle = (!flagship->IsThrusting() && !flagship->IsSteering()
+				&& !flagship->IsReversing() && flagship->Velocity().LengthSquared() < idleMaxSpeed);
+		if(requirePeaceful)
+			for(const Hardpoint &weapon : flagship->Weapons())
+				shipIdle &= !weapon.WasFiring();
```
I'm honestly not sure. We should have someone more familiar with the firing code check. However, even without it, it should just mean that if `requirePeaceful` is set and the ship fires, the reset is delayed by a single frame, shouldn't it?

TomGoodIdea replied:
> However, even without it, it should just mean that if `requirePeaceful` is set and the ship fires, the reset is delayed by a single frame, shouldn't it?

Yeah I think so ~~but it's still a single frame~~. Let's wait for another opinion.

TomGoodIdea commented:
Some of my recent review comments haven't been resolved yet.

danaris commented:
> Some of my recent review comments haven't been resolved yet.

Sorry about that, just hadn't pushed them!

Amazinite commented:
Heads up that I'm planning to do a final review of this some time in April so that we can finally get this feature in. (Sorry for disappearing, before. 👉👈 )

danaris commented:
> Heads up that I'm planning to do a final review of this some time in April so that we can finally get this feature in. (Sorry for disappearing, before. 👉👈 )

No worries; I get it very well. (I, too, am a volunteer lead dev/admin on a small indie project!)

TomGoodIdea commented:
A reminder that there are still a few not resolved review comments (and now some merge conflicts appeared).

Amazinite commented on source/Mission.cpp:
```diff
@@ -318,6 +318,8 @@ void Mission::Load(const DataNode &node, const ConditionsStore *playerConditions
 		else if(child.Token(0) == "npc")
 			npcs.emplace_back(child, playerConditions);
+		else if(child.Token(0) == "timer")
+			timers.emplace_back(child);
```
```suggestion
			timers.emplace_back(child, playerConditions);
```

Amazinite commented on source/Mission.cpp:
```diff
@@ -1548,6 +1565,10 @@ Mission Mission::Instantiate(const PlayerInfo &player, const shared_ptr<Ship> &b
 
+	// Instantiate the timers.
+	for(const Timer &timer : timers)
+		result.timers.push_back(timer.Instantiate(player.Conditions(), subs, sourceSystem, jumps, payload));
```
```suggestion
		result.timers.push_back(timer.Instantiate(subs, sourceSystem, jumps, payload));
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,387 @@
+
+
+
+Timer::Timer(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Timer::Load(const DataNode &node)
```
```suggestion
Timer::Timer(const DataNode &node, const ConditionsStore *playerConditions)
{
	Load(node, playerConditions);
}



void Timer::Load(const DataNode &node, const ConditionsStore *playerConditions)
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,387 @@
+		else if(child.Token(0) == "reset fired")
+			resetFired = true;
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
+			actions[TimerTrigger::TIMEUP].Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			actions[TimerTrigger::RESET].Load(child);
```
```suggestion
			actions[TimerTrigger::TIMEUP].Load(child, playerConditions);
		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
			actions[TimerTrigger::RESET].Load(child, playerConditions);
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,387 @@
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action.
+Timer Timer::Instantiate(const ConditionsStore &store, map<string, string> &subs, const System *origin,
+	int jumps, int64_t payload) const
```
```suggestion
Timer Timer::Instantiate(map<string, string> &subs, const System *origin, int jumps, int64_t payload) const
```

Amazinite commented on source/Timer.cpp:
```diff
@@ -0,0 +1,387 @@
+		return result;
+	}
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(store, subs, origin, jumps, payload);
```
```suggestion
		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+#include <map>
+#include <string>
+
+class DataNode;
```
```suggestion
class ConditionsStore;
class DataNode;
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+
+public:
+	Timer() = default;
+	Timer(const DataNode &node);
+	// Set up the timer from its data file node.
+	void Load(const DataNode &node);
```
```suggestion
	Timer(const DataNode &node, const ConditionsStore *playerConditions);
	// Set up the timer from its data file node.
	void Load(const DataNode &node, const ConditionsStore *playerConditions);
```

Amazinite commented on source/Timer.h:
```diff
@@ -0,0 +1,126 @@
+	bool IsComplete() const;
+
+	// Calculate the total time to wait, including any random value.
+	Timer Instantiate(const ConditionsStore &store, std::map<std::string, std::string> &subs, const System *origin,
+		int jumps, int64_t payload) const;
```
```suggestion
	Timer Instantiate(std::map<std::string, std::string> &subs, const System *origin, int jumps, int64_t payload) const;
```

Amazinite reviewed (commented):
I believe this is all the changes necessary to match the latest changes to actions.

Amazinite commented:
Thoughts on changing the proximity node to have the following behavior instead?
```
"proximity to" object [(<planet name> | "center" | <object-index#> | "any")]...
	location
		<location filter>
	...
	distance <distance#>
	(close | far)
"proximity to" npc [(<ship name> | "flagship" | <npc-index#> | "any")]...
	distance <distance#>
	(close | far)
```
The first format allows you to specify which objects in a system you need to be in proximity to. This can be a named planet, the center of the system, an integer specifying the index of the object in the system (allowing you to specify a particular object that isn't a named planet), "any" to say you need to be near any object at all, and/or a location filter.

The second format says you need to be near an NPC that is a part of the mission. You can name the NPC, use "flagship" to specify that you need to be near the flagship of an NPC block, provide an index to which NPC block you need to be near, and "any" to say you need to be near any NPC in the mission.

Each "proximity to" node could then children for the proximity settings, measuring distance and if you need to be close or far from the target. A single proximity node can specify multiple objects or NPCs to say that you need to be near one of any of them. A timer can have multiple proximity nodes to specify that you need to remain near multiple things at once.

Amazinite commented:
Continuing to evolve this ProximityFilter idea. 🤔 
```
"proximity to" ("stellar object" | "mission npc" | "system center")
	"any" # Any stellar object or mission NPC is enough
	# Planet name or NPC name
	"object name" <name>...
		<name>
		...
	# Object sprite or ship model
	"object type" <name>...
		<name>
		...
	# Index of the object in the system or NPC block in the mission
	"object index" <value#>...
		<value#>
		...
	# A location filter that matches the planet or mission NPC
	location
		<location filter>
	distance <distance#>
	(close | far)
```

danaris commented:
I like it—it does allow for several more interesting applications of the Timer idea.

I can see several places where we could replace an `on enter` trigger with a Timer trigger that requires being vaguely near a particular stellar or NPC, even if only for a few seconds—such as the trip to Sagittarius A* with Dawn. Actually, we could even do both: an `on enter` that says "OK, now hang out for a bit to get a reading", and then once you've stayed within visual range of the black hole for, say, 10 seconds, trigger the rest of the conversation.

Amazinite commented:
@danaris @TomGoodIdea I've reworked this PR to be a minimum viable product for mission timers. 

The syntax is now as follows:
```
timer
	time <base time#> [<random time#>]
	elapsed <time#> # Instantiated only.
	pauses
	optional
	peaceful
	(cloaked | uncloaked)
	solo
	idle [<speed#>]
	system <system name>
	system
		<location filter>
	on deactivation
		<action>
	on timeup
		<action>
```

For now, I've removed the reset conditions and the proximity nodes. I've made it so that if the timer deactivates, it always resets, unless the `pauses` tag is present, in which case it pauses.

The way that reset conditions were implemented still seemed too hard to read to me. In the future, I want to make is so that every call to `Deactivate` has its own enum that it passes into the function. If that enum is also present in a new "reset triggers" node, then that's what triggers a reset. 
```
"reset triggers"
	uncloaked
	cloaked
	"not idle"
	"not peaceful"
	"not solo"
	"not in system filter"
	"not in proximity filter"
	"changed system"
	"landed"
```
The last two things would be triggers that reset the timer, but don't deactivate it. Since the default behavior is already to reset if the requirements aren't met, these triggers would be used as a means of having a timer whose behavior is `pauses` still reset under certain conditions.

For proximity filters, see my earlier message about those.

Since the timer now only ever pauses or resets, I've renamed `on reset` to `on deactivation`.
I'd also like to have more `on *` triggers.

* `on elapsed <frames#>`: Trigger after a certain number of frames have elapsed (but before the timer has completed).
* `on activation`: Trigger once the timer starts ticking.
* `on deactivation [<start#> [<end#>]]`: Only trigger this deactivation action if at least a certain number of frames have elapsed (up to a higher number, so that you can have multiple deactivation actions that cover different ranges).
* `on reset [<start#> [<end#>]]`: A trigger that runs on deactivations that also reset the timer.
* `on pause [<start#> [<end#>]]`: Ditto, but for pausing.

And for actions that we want to be repeatable, I'd like to have a separate trigger name, such as `on "repeatable reset"`.

Amazinite commented:
I'm going to mode the `time` child node so that the base time and random time are inline with the `timer` root. A timer must have a time given to it, so it doesn't make much sense to me to have it be a child node that could be excluded.

TomGoodIdea commented on source/MissionTimer.cpp:
```diff
@@ -0,0 +1,386 @@
+		timeElapsed = 0;
+
+	// Perform the DEACTIVATION action, if there is one, assuming it hasn't fired yet.
+	if(!triggeredActions.contains(TimerTrigger::DEACTIVATION))
```
Let's not search the set twice.
```suggestion
	if(triggeredActions.insert(TimerTrigger::DEACTIVATION).second)
```

TomGoodIdea commented on source/MissionTimer.cpp:
```diff
@@ -0,0 +1,386 @@
+		auto it = actions.find(TimerTrigger::DEACTIVATION);
+		if(it != actions.end())
+			it->second.Do(player, ui, &mission);
+		triggeredActions.insert(TimerTrigger::DEACTIVATION);
```
```suggestion
```

TomGoodIdea commented:
Does anyone have a save that could be used to test the actual in-game data changes to Sad Archie?

TomGoodIdea commented:
@danaris considering we are back in the unstable phase, could you take a look at my review comments?

danaris commented:
I might be able to at some point, but @Amazinite did a pretty thorough rework, so I'd have to bring myself fully up to speed on what his changes did to it first. (Also it won't be till after next week, when I'll be out of the country.)

TomGoodIdea commented:
Do you plan to work on this anytime soon, or should we label it as orphaned?

TomGoodIdea commented on source/MissionTimer.h:
```diff
@@ -0,0 +1,119 @@
+class ConditionsStore;
+class DataNode;
+class DataWriter;
+class Mission;
```
```suggestion
class Mission;
class MissionAction;
```

Amazinite commented on source/MainPanel.cpp:
```diff
@@ -124,6 +124,9 @@ void MainPanel::Step()
 	engine.Step(isActive);
 
+	if(isActive)
+		player.StepMissionTimers(GetUI());
```
We don't want to step timers if the engine is paused. 
```suggestion
	if(isActive && !engine.IsPaused())
		player.StepMissionTimers(GetUI());
```

Amazinite commented on source/MissionTimer.h:
```diff
@@ -0,0 +1,120 @@
+	// If true, the timer pauses instead of resetting when deactivated.
+	bool pauses = false;
+
+	// Whether any of the activation requirements below are checked by this timer.
```
```suggestion
	// Whether any of the activation requirements below must be checked by this timer.
```

Amazinite commented on source/MissionTimer.cpp:
```diff
@@ -0,0 +1,392 @@
+			return false;
+		// And their ship's velocity must be below the max speed threshold.
+		if(flagship->Velocity().LengthSquared() < idleMaxSpeed)
+			return false;
```
Am I crazy or are these the exact opposite of what they should be?

```suggestion
		// The player can't be sending movement commands.
		if(flagship->IsThrusting() || flagship->IsSteering() || flagship->IsReversing())
			return false;
		// And their ship's velocity must be below the max speed threshold.
		if(flagship->Velocity().LengthSquared() > idleMaxSpeed)
			return false;
```

Amazinite commented on source/MissionTimer.cpp:
```diff
@@ -0,0 +1,386 @@
+			return false;
+		// And their ship's velocity must be below the max speed threshold.
+		if(flagship->Velocity().LengthSquared() < idleMaxSpeed)
+			return false;
```
Would that be any different than just giving a super high idle max speed?

Amazinite commented on source/MissionTimer.cpp:
```diff
@@ -0,0 +1,386 @@
+		// are only used defensively and automatically do not count against the
+		// player.
+		for(const Hardpoint &hardpoint : flagship->Weapons())
+			if(!hardpoint.IsSpecial() && hardpoint.WasFiring())
```
I'd be concerned about the order that things are being executed in. IsFiring will only be true this frame if ship movement has already occurred. Now I think it is the case that we do ship movement and then step timers, but it feels error-prone to be checking this frame instead of last frame for no perceptible benefit.

Amazinite commented on tests/integration/config/plugins/integration-tests/data/tests/tests_timers.txt:
```diff
@@ -0,0 +1,67 @@
+#		assert
+#			has "short timer: done"
+#			not "long timer: done"
+#			has "optional long timer: done"
```
Yeah, looks like that was it.

warp-core commented:
> Will wiki documentation be added?

I think that's what https://github.com/endless-sky/endless-sky-wiki/pull/145 is supposed to be.
